File: Algorithms.DataStructures/Python/4_Task_Stack_max/Task4.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>


//#define DEBUG

int num_queries;

int stack[100000];
int stack_top = -1;					// stack pointer
int max_element;					// max element on the stack

int curr_val_to_push;				// next value to stack

char ret_str[400000]; 

void pop();
void push(int a);
bool is_stack_empty();

/* Function returns true is it is push operation
	and then sets global variable curr_push
*/
bool is_push_input(char* input, int* num_to_push);

/* Function returns true if it is "max" input
	else - false
	*/
bool is_max_input(char* input);

/* Get max value on the stack */
int get_max_on_stack(void);

int main(int argc, char* argv[]) {
	
	char input[64];
	bool is_push = false, is_max = false;
	
	#ifdef DEBUG
		printf("\n Enter number of queries...\n");
	#endif
	
	// get number of queries
	fgets(input, sizeof(input), stdin);
	sscanf(input,"%d", &num_queries);
	
	if( num_queries <= 0) return 0;
	
	for ( int i = 0; i < num_queries; i++)
	{
		#ifdef DEBUG
			printf("\n Enter operation and value (ex pop or push 34) or max \n");
		#endif
		fgets(input, 64, stdin); //sizeof(input)
		
		is_max = is_max_input(input);
		
		if ( !is_max) {
			is_push = is_push_input(input, &curr_val_to_push);
			
			if ( is_push ) {			
				// if stack is empty, set max value
				if( is_stack_empty() ) max_element = curr_val_to_push;
				
				push(curr_val_to_push);
				
				//if current element > max_element then reset max value
				if ( curr_val_to_push > max_element) max_element = curr_val_to_push;
			}
			else {			//pop operation
				if( !is_stack_empty() ){
					if ( stack[stack_top] >= max_element)
					{
						pop();
						if( !is_stack_empty() ) {
							//reevaluate maximum element on the stack
							max_element = get_max_on_stack();
						}
					}
					else {
						//just pop
						pop();
					}
				}	
			}
			
		}
		else { //actions if max was entered
			//if stack is not empty - show current max value on the stack
			
		 	if( !is_stack_empty() ) {
				//olf #55test failed// printf("%d\n", max_element);
				char curr_max[60];
				sprintf(curr_max,"%d\n",max_element);
				strcat(ret_str,curr_max);
				//append ret string
			}
		}
		
	#ifdef DEBUG
		if( !is_max)
			printf("\n [%d] %s %d \n", i, (is_push ? "push" : "pop"), ( !is_push ? -1 : curr_val_to_push) );
		else
			printf("\n [%d] %s \n", i, "max");
	#endif
		
	}

	#ifdef DEBUG
	printf("%s\n\n\n", ret_str);
	#endif
	
	fwrite(ret_str, 1, strlen(ret_str), stdout);
	//printf("%s", ret_str);
	//puts(ret_str);

	return 0;
}

/* Function returns true if it is push operation
	and then sets global variable curr_push
*/
bool is_push_input(char* input, int* num_to_push)
{
	char operation_type[64]; //pop or push
	bool POP = false, PUSH = true;
	static char* pop_s = "pop";
	static char* push_s = "push";

	sscanf(input,"%s%d", operation_type,&curr_val_to_push);

	if(memcmp(operation_type, pop_s, 3) == 0) 
	{
		return POP;
	}
	else if (memcmp(operation_type, push_s, 4) == 0) 
	{
		//old sscanf(input,"%s%d", operation_type, &curr_val_to_push);//num_to_push);
		return PUSH;
	}
	return POP;	
}

/* Function returns true if it is "max" input
	else - false
	*/
bool is_max_input(char* input)
{
	if( memcmp(input, "max", 3) == 0) return true;
	
	return false;
}


void push(int a) 
{
	stack_top++;
	stack[stack_top] = a;
}

void pop() 
{
	if ( is_stack_empty()) return;
	stack[stack_top] = 0;
	stack_top--;
}

bool is_stack_empty(void) 
{

	if ( stack_top <= -1 ) {
		stack_top = -1;
		return true;
	}

	return false;
}

/* Get max value on the stack */
int get_max_on_stack(void) 
{
	int max_element;
	if ( stack_top == 0) {
		max_element = stack[0];
		return max_element;
	}
	max_element = stack[0];
	
	for ( int i = 0; i <= stack_top; i++)
	{
		if( stack[i] > max_element) max_element = stack[i];
	}
	return max_element;
}

File: Algorithms.DataStructures/Python/4_Task_Stack_max/Task4.py
import sys
import re				# regular expressions library

print('Hello!')

pyStack = [] 			# Stack instructions (pop, push, max...)
#pyStackNumb = [] 		# Stack numbers 0, 100, -2

stack_top=-1 			# points to the top of the Stack

retString=str() 		# resulted line (output)

curr_val_to_push=-1		# current value to push
max_element=-1			# current maximum value in the stack

DEBUG=1

''' 
Function returns true if there is "max" in tempStr
	else - false
'''
def is_max_input(tempStr):
	
	if "max" in tempStr:	#check if there is #max in tempStr
		return True
	else:
		return False

''' 
Function returns true if pyStack is empty
'''
def is_stack_empty(stack_top_loc):
	if (stack_top_loc <= -1):
		stack_top_loc = -1
		return True
	return False
	
'''  
	Function returns true if it is push operation
	and then sets global variable curr_push
'''
def is_push_input(tempInput,curr_val_to_push):
	POP = False
	PUSH = True
	pop_s = "pop"
	push_s = "push"
	
	if pop_s in tempInput:
		return POP
	elif push_s in tempInput:
		curr_val_to_push = int(re.search(r'\d+', tempInput).group())	#use regular expression lib to find an integer
		#print('re: ', curr_val_to_push) 
		return PUSH

'''
	Function pushes an element ont the stack
'''		
def push(element, stack_top_pointer): 
	stack_top_pointer +=1
	pyStackNumb[stack_top_pointer]=element
	print('local',pyStackNumb)
	return

'''
	Remove top element from the stack
'''
def pop(stack_top_local) :
	bIsStackEmpty = is_stack_empty(stack_top_local)
	
	#if stack is empty, there is no element to remove
	if bIsStackEmpty:
		return
	
	pyStackNumb[stack_top_local] = 0
	stack_top_local -= 1

'''
 Get max value on the stack
'''
def get_max_on_stack(stack_top_local):
	max_element=0	#declare variable
	if ( stack_top_local == 0):
		max_element = pyStackNumb[0];
		return max_element

	max_element = stack[0]
	
	##derive max element in the stack
	for idx in range(stack_top_local+1):
		if (pyStackNumb[idx] > max_element):
			max_element = pyStackNumb[idx]
	
	return max_element



### Start of the program ###
#init values once again


# Get number of queries
num_queries = int(input())

'''
if DEBUG:
	print('Queries: ', num_queries)
'''

#fool protection
if num_queries <= 0:
	print('Error 1')
	exit(0)
	
pyStackNumb = [num_queries] 		# Stack numbers 0, 100, -2
	

pyString=str()	#init empty string

##Get all commands from keyboard
for numQue in range(num_queries):
	tempStr=raw_input()
	pyString += tempStr
	pyString += '\n'			# add empty line

# DEBUG
#print(pyString)
#print(pyString.splitlines())

## Split inut. Assign command to every next index
pyStack = pyString.splitlines()

#Print prev operation result
'''
if DEBUG:
	lengthStack = len(pyStack)
	for index in range(lengthStack):
		print(pyStack[index])	
'''

## Going through the whole range of instructions
for queryIdx in range(num_queries):
	#print(queryIdx)
	bIsMax = is_max_input(pyStack[queryIdx]) # is there a max instruction?
	
	if bIsMax != True :
		bIsPush = is_push_input(pyStack[queryIdx],curr_val_to_push)
		if (bIsPush == True) :	#PUSH OPERATION
			print('PUSH!')
			# if stack is empty, set max value
			bIsStackEmpty = is_stack_empty(stack_top) 
			if (bIsStackEmpty == True):
				max_element = curr_val_to_push
			
			push(curr_val_to_push,stack_top)
			print('global',pyStackNumb)
			
			#if current element > max_element then reset max value
			if (curr_val_to_push > max_element):
				max_element = curr_val_to_push
		#POP OPERATION
		else:
			bIsStackEmpty = is_stack_empty(stack_top)
			if ( bIsStackEmpty != True):
				if (pyStackNumb[stack_top] > max_element):
					pop(stack_top)
					
					##reevaluate maximum element on the stack
					bIsStackEmpty = is_stack_empty(stack_top)
					if (bIsStackEmpty != False):
						max_element = get_max_on_stack(stack_top);
			else:
				pop(stack_top) #just pop		
	else:
		'''
		actions if max was entered
		if stack is not empty - show current max value on the stack
		'''
		print('MAXMUM')
		bIsStackEmpty = is_stack_empty(stack_top) 
		if (bIsStackEmpty != True): #if stack is not empty
			#append output string
			if (DEBUG==1):
				print('HERE!')
			retString += str(max_element)
			retString += '\n'

## Print output result!
if DEBUG==1:
	print('Result: ', retString)
else:
	print(retString)
	

